yi-core/src/Yi/CompletionTree.hs

```diff
@@ -1,59 +1,112 @@
+  inits       = T.inits
 
-obvious :: CompletionTree a -> ([a],CompletionTree a)
-obvious (CT [(letter,moretrees)]) = first ((:) letter) $
-                                             obvious moretrees
-obvious remainingchoice           = ([],remainingchoice)
+instance Eq a => Completable [a] where
+  stripPrefix = L.stripPrefix
+  isPrefixOf  = L.isPrefixOf
+  inits       = L.inits
 
-mergeTrees :: Ord a => [CompletionTree a] -> CompletionTree a
-mergeTrees a = mergeTrees' (map sort' a)
-  where sort' = CT . sortBy (compareBy fst) . (\(CT l)->l)
+-- | This function converts a list of completable elements to a CompletionTree
+-- It finds elements that share a common prefix and groups them.
+fromCompletable :: (Ord a, Completable a, Show a) => [a] -> CompletionTree a
+fromCompletable [] = mempty
+fromCompletable (x:xs)
+  | x == mempty = CT (Map.singleton mempty mempty) <> fromCompletable xs
+  | otherwise = case maximumBy' (compare `on` childrenIn xs) (tail $ inits x) of
+      Nothing -> CT (Map.singleton x mempty) <> fromCompletable xs
+      Just parent -> case first (x:) $ partition (parent `isPrefixOf`) xs of
+        ([_],rest) -> CT (Map.singleton parent mempty) <> fromCompletable rest
+        (hasParent, rest) -> CT (Map.singleton parent . fromCompletable $
+           map (fromJust . stripPrefix parent) hasParent) <> fromCompletable rest
+      -- A parent is the prefix and the children are the items with the parent as prefix
+      where childrenIn list parent = length $ filter (parent `isPrefixOf`) list
 
-mergeTrees':: Ord a => [CompletionTree a] -> CompletionTree a
-mergeTrees' trees = CT $
-                    map (\x->(fst $ head x,mergeTrees $ map snd x)) $
-                    groupBy (((EQ==).).compareBy fst)  $
-                    sortBy (compareBy fst) $
-                    concatMap (\(CT x)->x) trees
+-- | The largest element of a non-empty structure with respect to the
+-- given comparison function, Nothing if there are multiple 'largest' elements.
+maximumBy' :: Eq a => (a -> a -> Ordering) -> [a] -> Maybe a
+maximumBy' cmp l | atleast 2 (== max') l = Nothing
+                 | otherwise = Just max'
+  where  max' = maximumBy cmp l
+         -- This short-circuits if the condition is met n times before the end of the list.
+         atleast :: Int -> (a -> Bool) -> [a] -> Bool
+         atleast 0 _ _ = True
+         atleast _ _ [] = False
+         atleast n cmp' (x:xs) | cmp' x = atleast (n - 1) cmp' xs
+                               | otherwise = atleast n cmp' xs
 
-complete :: Eq a => CompletionTree a -> [a] -> ([a],CompletionTree a)
-complete tree    [] = ([],tree)
-complete (CT []) _  = ([],CT [])
-complete (CT level) (a:ta) = first ((:) a) $
-    case match of
-       Just m -> complete (snd m) ta
-       Nothing -> ([],CT [])
-     where match = find ((a==).fst) level
+-- | Complete as much as possible without guessing.
+obvious :: Completable a => CompletionTree a -> (a, CompletionTree a)
```

Comment:
I know this does what the old `obvious` did but I think `complete` is a better name, if my interpretation of the code is correct. The `(a, CompletionTree a)` has, I think, the following properties: `a` is the "obvious" which we were able to get without guessing, and `CompletionTree a` is the rest of the completions. The `complete` function used to give us an `[a]` but I believe we can now get this with what you called `toCompletable` but what I am calling `toList`.